frontend/src/components/ProjectCardGrid/index.tsx

```diff
@@ -0,0 +1,95 @@
+import React, { useEffect } from 'react'
+import { useInView } from 'react-intersection-observer'
+import { Filter } from 'meilisearch'
+
+import { meiliIndex } from '@utils/meili'
+import ProjectCard, { IntersectionObserverRef } from '../ProjectCard'
+import styles from './index.module.scss'
+
+export const cardsPerRow = 3
+const defaultLimit = cardsPerRow * 6
+
+export interface Project {
+  id: string
+  ownerName: string
+  projectName: string
+  repoName: string
+  summary: string
+}
+
+interface ProjectCardGridProps {
+  projects: Project[]
+  intersectionObserverRef?: IntersectionObserverRef
+}
+
+const ProjectCardGrid = ({
+  projects,
+  intersectionObserverRef,
+}: ProjectCardGridProps) => {
+  return (
+    <div className={styles.cardsGrid} data-cy="cards-grid">
+      {projects?.map((project, index) => (
+        <ProjectCard
+          {...project}
+          key={project.id}
+          intersectionObserverRef={
+            index === projects.length - cardsPerRow * 2
+              ? intersectionObserverRef
+              : null
+          }
+        />
+      ))}
+    </div>
+  )
+}
+
+interface SearchSWRKeyParams {
+  filter: Filter
+  limit: number
+  offset: number
+  query: string
+}
+
+export const getKey =
+  ({ query, filter }: Partial<SearchSWRKeyParams>) =>
+  (pageIndex: number, previousPageData: Project[]): SearchSWRKeyParams => {
+    if (previousPageData && !previousPageData.length) {
+      // reached the end
+      return null
+    }
+    return { query, offset: pageIndex * defaultLimit, limit: defaultLimit, filter }
+  }
+
+export type GridFetcherArgs = Partial<SearchSWRKeyParams>
+
+export const gridFetcher = async ({
+  filter,
+  offset = 0,
+  query = '',
+  limit = defaultLimit,
+}: GridFetcherArgs) => {
+  const searchResult = await meiliIndex.search(query, { limit, offset, filter })
+  return searchResult.hits as Project[]
+}
+
+type SetSWRInfiniteSize = (
+  size: (_size: number) => number,
+) => Promise<Project[][] | undefined>
+
+/**
+ * Update useSWRInfinite size when the user reaches two rows from the end of the grid.
+ * @param  setSize - the function returned by useSWRInfinite.
+ */
+export const useUpdateBeforeReachingLimit = (setSize: SetSWRInfiniteSize) => {
+  const [ref, isReachingLimit] = useInView({ triggerOnce: true })
+
+  useEffect(() => {
+    if (isReachingLimit) {
+      setSize(size => size + 1)
+    }
+  }, [isReachingLimit, setSize])
+
+  return ref
+}
```

Comment:
I'm wondering about this. The name `useUpdateBeforeReachingLimit` is pretty intimidating and I don't like that `setSize` is passed in like that. Would this be clearer if `useSWRInfinite` and `useInView` were wrapped up in one hook?

```ts
export interface SearchParams {
  query: string
  filter: string
}
export const useSearch = (params: SearchParams) => {
  const [ref, isReachingLimit] = useInView({ triggerOnce: true })
  const { data, setSize } = useSWRInfinite(getKey(params), gridFetcher)

  useEffect(() => {
    if (isReachingLimit) {
      setSize(size => size + 1)
    }
  }, [isReachingLimit, setSize])

  return { projects: data?.flat(), ref }
}
```

Reply:
Resolved in 11b9354.